### ISSUE TITLE
Split up `common_methods_invocations.py`?

### ISSUE THREAD

**Issue by peterbell10** [pytorch/pytorch#82518]
`common_methods_invocations.py` has grown to 22K lines and over 1MB in file size. One implication of this is you can't open it in the GitHub UI or link to specific lines of code.

I propose creating an opinfos folder and where there are currently different categories, such as `UnaryUfuncInfo` or `ReductionOpInfo`, these could be their own file.

**Comment by Lezcano** [comment]
fwiw, @antocuni had an interesting proposal for this, and I think he even had an open PR, but we never merged it (I don't have the PR at hand now, but it should be easy to find)

**Comment by antocuni** [comment]
I think you are referring this https://github.com/pytorch/pytorch/issues/55159 and https://github.com/pytorch/pytorch/pull/59871, which unfortunately never went anywhere

**Comment by albanD** [comment]
Not being able to see it on GitHub is a pain for sure.
One thing that was mentioned before though is that there is a benefit to having all of these in a single file as it makes doing a ctrl+f to find functions much easier.

cc @jbschlosser @zou3519 @ngimel @kshitij12345 for opinions!

**Comment by zou3519** [comment]
Mass modification of OpInfos is difficult. Splitting up the files would make mass modification of OpInfos even more difficult. I'm not opposed to splitting up the files; as part of the work of splitting up the files we have the opportunity to make the situation better.

Here's what's difficult about modifying OpInfos today:
- Let's say I add a new test, and want to add skips for 20 different OpInfos. Then finding the OpInfo and then pasting in the skip takes a while, especially because (1) the indentation between OpInfos is not consistent and (2) OpInfos don't have consistent formatting -- the skips/decorators field is not required. A potential option here is to enforce a style guide and autoformatter for OpInfos
- Finding an OpInfo is difficult. Need the OpInfo for linear? grep for "linear" -- this will get you sample_inputs_linear, the sample inputs for bilinear, the OpInfo for bilinear, the OpInfo for linear, and some OpInfos for interpolate (with mode = bilinear). I don't know how to make this better, so the question here is: why do people need to look for OpInfos? The reason why I look for OpInfos is to inject skips + xfail decorators: maybe we should move towards putting the skips/decorators on the tests rather than on the OpInfo itself (we already do this in functorch -- https://github.com/pytorch/pytorch/blob/42fefd4403829406427b0e2912d28290d60ef86a/functorch/test/test_ops.py#L298-L306)

**Comment by peterbell10** [comment]
Another alternative would be to split the opinfos by torch module:
```
opinfos/linalg.py
opinfos/fft.py
opinfos/nn.py
opinfos/special.py
opinfos/torch.py
```

This way you can tell which file it should be in without the need to `git grep`.

**Comment by jbschlosser** [comment]
> Another alternative would be to split the opinfos by torch module:

I like this a lot and am leaning towards something like this for some magical future where I finally get around to splitting `test_nn.py`.

**Comment by peterbell10** [comment]
> For someone writing a new op, I expect [declaring skips in OpInfo] to be somewhat more convenient. For test writers / maintainers / feature developers, I imagine [declaring skips on the test] would make patterns of failure clearer. 

Another consideration is which side of the equation needs action to get the test to pass? If an operator skips several tests due to a  bug, then fixing the bug and clearing those skips in one go is much easier from the OpInfo. On the other hand if a feature fails for structural reasons down steam of the operator itself, then we fix that structural issue and clear skips for several operators at a time.

**Comment by jbschlosser** [comment]
> Another consideration is which side of the equation needs action to get the test to pass? If an operator skips several tests due to a bug, then fixing the bug and clearing those skips in one go is much easier from the OpInfo. On the other hand if a feature fails for structural reasons down steam of the operator itself, then we fix that structural issue and clear skips for several operators at a time.

This is true, and I'll put forth the idea that optimizing for feature development over individual op development is more useful in practice due to the scope difference between the two. I *hope* it's unlikely that a dev commits known buggy op logic with several skips to pass CI. OTOH, it's common for features to be developed incrementally since they require much more work, and exFails are useful here to pass CI for parts of the feature that have not been completed / fixed yet.

**Comment by peterbell10** [comment]
> I hope it's unlikely that a dev commits known buggy op logic with several skips to pass CI. 

Of around 1300 total `DecorateInfo` instances:
- `test_variant_consistency` matches 235 lines
- `test_.*reference` matches 217 lines
- `test_out` matches 85 lines

so more than half of decorators are likely issues with the operator itself.

**Comment by Lezcano** [comment]
From the perspective of a person that has done a fair amount of work on maintaining some operators (namely `linalg` but also others), it's very useful to have a comprehensive list of all the tests that a given operation does not pass, as you may target them individually when you're maintaining that op.

That being said, I completely understand Richard's point as well. In my opinion, I think we could go with one of the options, and then just have a tool that given that option, it transforms it to the other one.
For example, if we go with listing all the ops failing test X in one place, this tool would allow you to compile, given a particular OpInfo, all the tests that it fails (and even better, where are the skip lists for these tests). That way, you can target them individually.
